### PR TITLE
Fix all_gather_cat producing wrong shapes on single-device runs

### LINKED ISSUE #421: Bug: all_gather_cat silently corrupts metric shapes and values on single-device runs
## Description

`ARModel.all_gather_cat` unconditionally calls `.flatten(0, 1)` on the result of `self.all_gather()`. On multi-GPU (DDP), `all_gather` prepends a new dimension `(K, d1, d2, ...)`, so flattening dims 0 and 1 correctly produces `(K*d1, d2, ...)`. However, on a single device, `all_gather` returns the tensor **unchanged** — no new leading dimension is added. This means `flatten(0, 1)` **merges the first two existing dimensions**, silently producing wrong tensor shapes and therefore wrong metric values.

## Current behavior

On single GPU/CPU, in `aggregate_and_plot_metrics`:

```
Input to all_gather_cat:  (N_eval, pred_steps, d_f)
After all_gather:         (N_eval, pred_steps, d_f)  ← no new dim on single device
After flatten(0, 1):      (N_eval * pred_steps, d_f)  ← WRONG: merged batch and time dims
After mean(dim=0):         (d_f,)                      ← WRONG: should be (pred_steps, d_f)
```

This means:
- **Validation/test metrics are silently averaged across both samples AND time steps**, instead of just across samples
- Downstream `plot_error_map` receives a 1D `(d_f,)` tensor instead of 2D `(pred_steps, d_f)`, causing a crash (partially patched by #193 in `vis.py`, but the values remain wrong)
- `on_test_epoch_end` spatial loss maps similarly have corrupted shapes: `(N_test * N_log, num_grid_nodes)` instead of `(N_test, N_log, num_grid_nodes)`, giving nonsensical spatial loss averages

## Expected behavior

On single device, `all_gather_cat` should return the tensor unchanged since there is nothing to gather/concatenate from other ranks.

## Reproduction

```python
import torch
import pytorch_lightning as pl

# Simulate single-device all_gather behavior:
tensor = torch.randn(4, 3, 5)  # (N_eval, pred_steps, d_f)
gathered = tensor  # SingleDeviceStrategy.all_gather returns tensor unchanged
result = gathered.flatten(0, 1)  # BUG: (12, 5) instead of (4, 3, 5)
averaged = result.mean(dim=0)   # (5,) instead of (3, 5)

print(f"Input: {tensor.shape}")
print(f"After flatten(0,1): {result.shape}")     # (12, 5) ← WRONG
print(f"After mean(dim=0): {averaged.shape}")     # (5,)   ← WRONG
```

## Root cause

[`ar_model.py` line 331](https://github.com/mllam/neural-lam/blob/main/neural_lam/models/ar_model.py#L331):
```python
def all_gather_cat(self, tensor_to_gather):
    return self.all_gather(tensor_to_gather).flatten(0, 1)
```

PyTorch Lightning's `SingleDeviceStrategy.all_gather` simply returns the tensor unchanged ([source](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/strategies/single_device.py)):
```python
def all_gather(self, tensor, group=None, sync_grads=False):
    return tensor
```

## Proposed fix

Add a dimension guard before flattening:

```python
def all_gather_cat(self, tensor_to_gather):
    gathered = self.all_gather(tensor_to_gather)
    if gathered.dim() > tensor_to_gather.dim():
        return gathered.flatten(0, 1)
    return gathered
```

## Affected code paths

- `aggregate_and_plot_metrics` (line 647): validation and test metric aggregation
- `on_test_epoch_end` (line 700): spatial loss map aggregation

## Related

- #192 / PR #193 — patches the downstream crash in `vis.py` but does **not** fix the root cause (metrics remain silently wrong)
- PR #208 — includes this fix bundled in a large refactor, but that PR is not yet merged

## Impact

**High** — Any single-GPU training or evaluation silently produces wrong validation/test metrics. Users get metrics that are averaged across both time steps and samples, giving misleadingly low error values. The test suite itself acknowledges this with the comment:
```python
# XXX: `devices` has to be set to 2 otherwise
# neural_lam.models.ar_model.ARModel.aggregate_and_plot_metrics fails
# because it expects to aggregate over multiple devices
devices=2,
```

### PR DESCRIPTION
**Pull request by RajdeepKushwaha5**
## Describe your changes

`ARModel.all_gather_cat` unconditionally calls `.flatten(0, 1)` on the result of `self.all_gather()`. On multi-device runs this is correct because `all_gather` stacks tensors along a new leading dimension `(K, d1, d2, …)`, and flattening merges it back to `(K*d1, d2, …)`.

On **single-device** runs (the default `SingleDeviceStrategy`), `all_gather` returns the tensor **unchanged** — no new dimension is added. The unconditional `.flatten(0, 1)` then merges the first two *real* dimensions of the data, silently producing wrong shapes and corrupted metrics.

**Example:**
```
Input tensor:             (N_eval, pred_steps, d_f) = (4, 3, 5)
Single-device all_gather: (4, 3, 5)          # unchanged
flatten(0, 1):            (12, 5)            # WRONG — merged N_eval × pred_steps
```

The fix adds a dimension guard — only flatten when `all_gather` actually introduced a new leading dimension:
```python
gathered = self.all_gather(tensor_to_gather)
if gathered.dim() > tensor_to_gather.dim():
    return gathered.flatten(0, 1)
return gathered
```

No new dependencies required.

## Issue Link

Closes #421

## Type of change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [x] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by RajdeepKushwaha5** [comment]
> Good change, code looks good to me and the tests are suitable. Could you:
> 
> 1. Check that all the fields in the PR template are correctly there. I think some checkboxes for assignees are missing.
> 2. Add a changelog entry for this.

Thanks for the review, Sir! I've pushed a commit with the CHANGELOG entry (under ### Fixed in [unreleased]), and updated the PR description to include the full assignee checklist from the template.
Let me know if anything else needs adjusting.
